### PR TITLE
[StateHandling] State API for remote platforms 

### PR DESCRIPTION
**Pull request by 1tnguyen**
### Description

State handling for remote platforms:

- Introduce `SimulationState` subclass for remote platform (`RemoteSimulationState` and `PyRemoteSimulationState`) which is essentially a holder of the kernel IR.

- Perform a lazy evaluation to resolve the state:
    
    (1) For state accessor, e.g., `amplitudes`: adding a configurable qubit number threshold. If the number of qubits is less than the threshold, resolve the full state vector to the client. Otherwise, send on amplitude access request. Adding an API to `cudaq::state` for batched amplitude access.
    
    (2) For overlap, sending job requests with both kernel IRs.
    
     
**Note**: the remote state is implemented for MLIR kernels only (C++ MLIR mode, kernel builder and Python).
Library mode C++ kernel is evaluated as-is (v0.7 mode): state is resolved to a vector by the server for each `get_state` call.

### CODE REVIEW

File: python/runtime/cudaq/algorithms/py_state.cpp
Comment (schweitzpgi) [inline python/runtime/cudaq/algorithms/py_state.cpp:10]: nit: remove blank lines between includes

File: python/runtime/cudaq/algorithms/py_state.cpp
Comment (schweitzpgi) [inline python/runtime/cudaq/algorithms/py_state.cpp:24]: Should this `tuple` have a name?

File: runtime/common/SimulationState.h
Comment (schweitzpgi) [inline runtime/common/SimulationState.h:10]: nit: add blank line between #pragma and #includes. :)

File: targettests/Remote-Sim/get_state.cpp
Comment (schweitzpgi) [inline targettests/Remote-Sim/get_state.cpp:33]: nit: delete this block?

File: runtime/cudaq/algorithms/get_state.h
Comment (schweitzpgi) [inline runtime/cudaq/algorithms/get_state.h:106]:
```suggestion
#elif defined(CUDAQ_REMOTE_SIM)
```

File: runtime/cudaq/algorithms/get_state.h
Comment (schweitzpgi) [inline runtime/cudaq/algorithms/get_state.h:112]:
```suggestion
```

File: runtime/cudaq/algorithms/get_state.h
Comment (schweitzpgi) [inline runtime/cudaq/algorithms/get_state.h:20]: nit: remove blank lines between includes

**Review by schweitzpgi — APPROVED** [review]
LGTM. You'll need additional approval though.

File: python/runtime/cudaq/algorithms/py_state.cpp
Comment (1tnguyen) [reply python/runtime/cudaq/algorithms/py_state.cpp]: It's a good idea. Added a struct in https://github.com/NVIDIA/cuda-quantum/pull/1653/commits/c497846919dbe40b66fae307afac3483794479f6 to name this tuple.

File: runtime/cudaq/algorithms/get_state.h
Comment (1tnguyen) [reply runtime/cudaq/algorithms/get_state.h]: Fixed.

File: targettests/Remote-Sim/get_state.cpp
Comment (1tnguyen) [reply targettests/Remote-Sim/get_state.cpp:33]: We'll keep this block since I haven't had a chance to retest it according to the ticket linked. We probably want to re-enable the test if it's no longer unstable.

**Comment by 1tnguyen** [comment]
Since this would introduce a breaking change to NVQC, it would be rebased to target `main` branch after the experimental branch is merged.

File: .github/workflows/integration_tests.yml
Comment (bmhowe23) [inline .github/workflows/integration_tests.yml:187]: I forget - was this change going to get pulled into a separate PR? Or was it just that we were going to hold off on this full PR until closer to the release.

Comment (1tnguyen) [reply .github/workflows/integration_tests.yml:187]: We will close this PR and migrate the changes to a new branch based on `main` with deployment to test the integration. The changes will be merged closer to release.

**Pull request closed without merging**
